libraries/moon/src/component-tests.js

```diff
@@ -0,0 +1,283 @@
+/**
+ * @license
+ * Copyright 2017 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Moon from 'moonjs';
+import {
+  ComponentWithoutChildren,
+  ComponentWithChildren,
+  ComponentWithChildrenRerender,
+  ComponentWithDifferentViews,
+  ComponentWithProperties,
+  ComponentWithUnregistered,
+  ComponentWithImperativeEvent,
+  ComponentWithDeclarativeEvent
+} from './components';
+import expect from 'expect';
+
+// Setup the test harness. This will get cleaned out with every test.
+let app = document.createElement('div');
+app.id = 'app';
+document.body.appendChild(app);
+let scratch; // This will hold the actual element under test.
+
+beforeEach(function() {
+  scratch = document.createElement('div');
+  scratch.id = 'scratch';
+  app.appendChild(scratch);
+});
+
+afterEach(function() {
+  app.innerHTML = '';
+  scratch = null;
+});
+
+describe('no children', function() {
+  it('can display a Custom Element with no children', function() {
+    let app = new ComponentWithoutChildren();
+    app.mount(scratch);
+    let root = app.el;
+    let wc = root.querySelector('#wc');
+    expect(wc).toExist();
+  });
+});
+
+describe('with children', function() {
+  function expectHasChildren(wc) {
+    expect(wc).toExist();
+    let shadowRoot = wc.shadowRoot;
+    let heading = shadowRoot.querySelector('h1');
+    expect(heading).toExist();
+    expect(heading.textContent).toEqual('Test h1');
+    let paragraph = shadowRoot.querySelector('p');
+    expect(paragraph).toExist();
+    expect(paragraph.textContent).toEqual('Test p');
+  }
+
+  it('can display a Custom Element with children in a Shadow Root', function() {
+    let app = new ComponentWithChildren();
+    app.mount(scratch);
+    let root = app.el;
+    let wc = root.querySelector('#wc');
+    expectHasChildren(wc);
+  });
+
+  it('can display a Custom Element with children in a Shadow Root and pass in Light DOM children', function() {
+    let app = new ComponentWithChildrenRerender();
+    app.mount(scratch);
+    let root = app.el;
+    let wc = root.querySelector('#wc');
+
+    Moon.nextTick(() => {
```

Comment:
I think any tests that use `Moon.nextTick` will need to add Mocha's [`done()` callback](https://mochajs.org/#asynchronous-code). Otherwise I noticed there were some errors showing up in the results. In the Vue tests I use `await` but since `Moon.nextTick` doesn't return a promise (at least I don't think it does?) we'll need to use `done()`.

This particular tests is interesting because if you put a `debugger` statement after line 85, and set `singleRun: false` in the karma.conf then you can debug the test in the browser and see that the value is still `1`. Which causes the expect at line 86 to fail. But if you wait an additional tick it increments to `2`. I think it's because the component waits a tick to call `this.set` which I assume occurs on the next tick. But the test needs to then wait two ticks. I'm not sure what the most elegant way to do this in Moon is, but we could just stack up two nextTicks in the tests 😄

Reply:
Yeah, for Moon's tests I have a custom `wait` function that runs any assertions and returns a Promise (which is then returned in the callback). Can definitely apply that here.

```js
const wait = function(cb) {
  return new Promise((resolve, reject) => {
    Moon.nextTick(() => {
      try {
        cb();
        resolve();
      } catch(err) {
        reject(err);
      }
    });
  });
}

// Usage
it("should do something", () => {
  // something async
  return wait(() => {
    expect(element.innerHTML).toEqual("foo");
  });
});
```

Reply:
Found the issue for the second thing you found. This seems to be because of an extra `Moon.nextTick` in the `ComponentWithChildrenRerender` component's `mounted` hook. The `mounted` hook is called _after_ the DOM is updated. Removing that extra `Moon.nextTick` will work.